src/proto/media/disk.rs

```diff
@@ -0,0 +1,206 @@
+/// protocol to a more general offset-length protocol. Firmware is
+/// reponsible for adding this protocol to any block I/O interface that
+/// appears in the system that does not already have a disk I/O protocol.
+#[repr(C)]
+#[unsafe_guid("ce345171-ba0b-11d2-8e4f-00a0c969723b")]
+#[derive(Protocol)]
+pub struct DiskIo {
+    revision: u64,
+    read_disk: extern "efiapi" fn(
+        this: &DiskIo,
+        media_id: u32,
+        offset: u64,
+        len: usize,
+        buffer: *mut u8,
+    ) -> Status,
+    write_disk: extern "efiapi" fn(
+        this: &mut DiskIo,
+        media_id: u32,
+        offset: u64,
+        len: usize,
+        buffer: *const u8,
+    ) -> Status,
+}
+
+impl DiskIo {
+    /// Reads bytes from the disk device.
+    ///
+    /// # Arguments:
+    /// * `media_id` - ID of the medium to be read.
+    /// * `offset` - Starting byte offset on the logical block I/O device to read from.
+    /// * `buffer` - Pointer to a buffer to read into.
+    ///
+    /// # Errors:
+    /// * `uefi::status::INVALID_PARAMETER` The read request contains device addresses that
+    ///                                     are not valid for the device.
+    /// * `uefi::status::DEVICE_ERROR`      The device reported an error while performing
+    ///                                     the read operation.
+    /// * `uefi::status::NO_MEDIA`          There is no medium in the device.
+    /// * `uefi::status::MEDIA_CHANGED`     `media_id` is not for the current medium.
+    pub fn read_disk(&self, media_id: u32, offset: u64, buffer: &mut [u8]) -> Result {
+        (self.read_disk)(self, media_id, offset, buffer.len(), buffer.as_mut_ptr()).into()
+    }
+
+    /// Writes bytes to the disk device.
+    ///
+    /// # Arguments:
+    /// * `media_id` - ID of the medium to be written.
+    /// * `offset` - Starting byte offset on the logical block I/O device to write to.
+    /// * `buffer` - Pointer to a buffer to write from.
+    ///
+    /// # Errors:
+    /// * `uefi::status::INVALID_PARAMETER` The write request contains device addresses that
+    ///                                     are not valid for the device.
+    /// * `uefi::status::DEVICE_ERROR`      The device reported an error while performing
+    ///                                     the write operation.
+    /// * `uefi::status::NO_MEDIA`          There is no medium in the device.
+    /// * `uefi::status::MEDIA_CHANGED`     `media_id` is not for the current medium.
+    /// * `uefi::status::WRITE_PROTECTED`   The device cannot be written to.
+    pub fn write_disk(&mut self, media_id: u32, offset: u64, buffer: &[u8]) -> Result {
+        (self.write_disk)(self, media_id, offset, buffer.len(), buffer.as_ptr()).into()
+    }
+}
+
+/// Asynchronous transaction token for disk I/O 2 operations.
+#[repr(C)]
+pub struct DiskIo2Token {
+    /// Event to be signalled when an asynchronous disk I/O operation completes.
+    pub event: Option<Event>,
```

Comment:
In trying to track down why the `Event` never gets signaled, I realized that this struct does not currently have the correct layout. We aren't using `NonNull` in `Event` like we do with `Handle`, so `Option<Event>` doesn't get the same layout as a raw pointer.

I filed https://github.com/rust-osdev/uefi-rs/issues/477 to fix this, but for now this line should be changed to `pub event: Event`.

Reply:
I think in this case this actually isn't necessary because I interpreted the specification wrong. The spec talks about the token "field" to be nullable, but it seems what it means is that the token argument to `ReadDiskEx`/`WriteDiskEx` can be `NULL` - meaning I could simply change the parameter type.